2021-09-30-review-apps-github-labels.md

```diff
@@ -0,0 +1,99 @@
+- Check that the user doing the action is an admin/maintainer/contributor of the repo so random people won't be able to trigger Review Apps creation adding labels (so this is safe to use in OSS projects)
+- Get all the info needed from the PR and the repo's secret (for security) to generate the payload for Heroku's API
+- Trigger the Review App concrete action (create/update/destroy)
+- Remove the label that triggered this action using Github's API
+
+### Resources
+
+We've found some resources and existing GitHub Actions to use as reference for most of the code needed:
+
+- Create Heroku's review app https://github.com/mheap/github-action-pr-heroku-review-app (this includes using the npm module, checking the user permissions, preparing the payload, hitting Heroku's `create` endpoint for review apps)
+- Add labels to a PR https://github.com/TimonVS/pr-labeler-action (this includes connecting to the GitHub API to set labels for a PR, we can use it as reference for removing the label instead)
+- How to create Github Actions https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action (this shows the step-by-step on creating github actions)
+- Documentation on setting ENV variables for the GitHub Actions https://docs.github.com/en/actions/learn-github-actions/environment-variables (needed for the Heroku's API token and Pipeline ID)
+
+## Complete workflow:
+
+- A developer creates a PR (no review app is created)
+- When the PR is ready for QA and the person doing the QA is ready, they will add the `deploy-review-app` label to the PR
+- After a moment, the GitHub Action will be triggered, creating a Review App and removing the `deploy-review-app` label (the information of the deploy should be added to the PR by Heroku like it's currently doing)
+- When the QA is done, the QA person will add the `destroy-review-app` label to the PR
+- After a moment, the GitHub Action will be triggered, destroying the Review App and removing the `destroy-review-app` label
+
+### Extra steps:
+
+- We could have a workflow configuration that destroys the review app when a `ready-to-merge` label is added (so the QA person don't need to remember to add the `destroy-review-app` label in that case).
+- We should have a workflow configuration that destroys the review app when the PR is closes (either close or merge+close) to prevent stale review apps running when not needed anymore.
+
+In these 2 cases, we would re-use the previous GitHub Action that destroy the Review App by just setting other jobs for the specific triggers. Here's an example of a GitHub Workflows job when a PR is closed https://github.com/arielj/test-gh-labels/blob/main/.github/workflows/on_closed.yml.
+
+# Final thoughts:
+
+The code of the GitHub Actions will be public and the different projects will reference them, so it's reusable. The only code that would be copied on each project is the workflows `.yml` file that sets the triggers.
```

Comment:
Does this mean there would be a repo for these actions and the implementing project just needs a reference to the repo with the GithubActions "code"?

Reply:
That would be cool because then we could open source the action itself!

Reply:
> Does this mean there would be a repo for these actions and the implementing project just needs a reference to the repo with the GithubActions "code"?

yes, that's how actions are created, you create a public repo and just add `uses: fastruby/create-review-app-from-label@v1` or whatever and github will get that repo and run the code in the job

like this https://github.com/mheap/github-action-pr-heroku-review-app